gh-101765: Fix SystemError / segmentation fault in iter `__reduce__` when internal access of `builtins.__dict__` exhausts the iterator

ionite34 opened the pull request:
# Summary
- Fixes potential segmentation fault or SystemError when `__reduce__` is called on `iter` objects when the key of the hash of `"iter"` within `__builtins__.__dict__` is a custom object that executes arbitrary code within `__eq__`, mutating the `iter` object and causing illegal memory access or SystemError (based on C argument evaluation order, which is undefined behavior).

## Affected methods
* `iterobject.c`
  - [x] `iter_reduce`
  - [x] `calliter_reduce`
* `listobject.c`
  - [x] `listiter_reduce_general`
* `bytearrayobject.c`
  - [x] `bytearrayiter_reduce`
* `bytesobject.c`
  - [x] `striter_reduce`
* `tupleobject.c`
  - [x] `tupleiter_reduce`
* `genericaliasobject.c`
  - [x] `ga_iter_reduce`

This PR also fixes a compounded issue where currently `genericaliasobject.ga_iter_reduce` does not handle empty iterators at all and has no NULL check
```python
Python 3.11.0 (main, Oct 26 2022, 10:14:06) [GCC 9.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> x = tuple[int]
>>>
>>> it = iter(x)
>>> _ = list(it)
>>>
>>> it.__reduce__()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
SystemError: NULL object passed to Py_BuildValue
```
Along with moving the evaluation of `_PyEval_GetBuiltin` for potential side effects, this also adds handling of the NULL case (like the other `iter_reduce` functions have:
```python
Python 3.12.0a5+ (heads/fix-reduce-dirty:93854e172e, Feb 10 2023, 13:41:45) [GCC 9.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> x = tuple[int]
>>>
>>> it = iter(x)
>>> _ = list(it)
>>>
>>> it.__reduce__()
(<built-in function iter>, ((),))
```

## Linked Issue
<!-- gh-issue-number: gh-101765 -->
* Issue: gh-101765
<!-- /gh-issue-number -->

ionite34 committed:
Fix undefined behavior in listiter_reduce from _PyEval_GetBuiltin side effects

ionite34 committed:
Update comment to not mention `__eq__`

ionite34 committed:
Fix undefined behavior in iter_reduce and calliter_reduce

ionite34 committed:
Update listiter_reduce_general comment

ionite34 committed:
Fix undefined behavior in bytearrayiter_reduce from _PyEval_GetBuiltin side effects

ionite34 committed:
Fix undefined behavior in striter_reduce from _PyEval_GetBuiltin side effects

ionite34 committed:
Fix undefined behavior in tupleiter_reduce from _PyEval_GetBuiltin side effects

JelleZijlstra commented on Objects/bytearrayobject.c:
```diff
@@ -2391,11 +2391,17 @@ PyDoc_STRVAR(length_hint_doc,
+
+    /* _PyEval_GetBuiltin can invoke arbitrary code.
+     * calls must be *before* access of `it` pointers,
+     * since C/C++ parameter eval order is undefined.
```
```suggestion
     * since C parameter eval order is undefined.
```
This is C code, so we don't care about C++.

ionite34 replied:
removed C++ references in 
[45522c6](https://github.com/python/cpython/pull/101769/commits/45522c6dd8f6d39b3aeb05015371e31a88168d29)

JelleZijlstra commented on Objects/listobject.c:
```diff
@@ -3443,19 +3443,26 @@ static PyObject *
 listiter_reduce_general(void *_it, int forward)
 {
     PyObject *list;
+    PyObject *iter;
```
You can just declare these within the block where they are used.

ionite34 replied:
Moved the declarations to the if blocks in [049a8dd](https://github.com/python/cpython/pull/101769/commits/049a8dda075a4b44cc7bd6997b2c42b70fead05e)

ionite34 committed:
Move iter call in unicodeiter_reduce before `it` pointer access due to potential side effects

ionite34 committed:
Add iter reduce tests for issue #101765

ionite34 committed:
Remove C++ reference in comments

ionite34 committed:
Remove C++ reference in comments

ionite34 committed:
Move builtin declarations inside if blocks

ionite34 committed:
Move _PyEval_GetBuiltin before gi checks, add gi NULL check in ga_iter_reduce

ionite34 committed:
Update iter reduce mutating tests for generic alias

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

ionite34 committed:
Fix backticks format for news

carljm commented on Lib/test/test_iter.py:
```diff
@@ -238,6 +244,52 @@ def test_mutating_seq_class_exhausted_iter(self):
+        # depending on the order of C argument evaluation, which is undefined
+
+        # Backup builtins.iter
+        builtins = __builtins__.__dict__ if hasattr(__builtins__, "__dict__") else __builtins__
```
```suggestion
        builtins = getattr(__builtins__, "__dict__", __builtins__)
```

in what cases (that need to be supported here) does `__builtins__` not have a `__dict__`?

ionite34 replied:
https://docs.python.org/3/reference/executionmodel.html?highlight=__builtins__#builtins-and-restricted-execution `__builtins__` is already a dict when not in the `__main__` module. I was thinking we may have both cases but I suppose we tests only run with the non `__main__` case, or in other words we only need `builtins = __builtins__`?

ionite34 replied:
Switched to `builtins = __builtins__` in [178b8ea](https://github.com/python/cpython/pull/101769/commits/178b8ea41bb045eed63d06674e2f337d4de97bef) since I don't see how unittest can be a `__main__` module

carljm replied:
The simplest and clearest approach might be to `import builtins` as recommended in the docs; then it will always be a module.

ionite34 replied:
changed to `import builtins` in [98ec3c6](https://github.com/python/cpython/pull/101769/commits/98ec3c63a973d06e784f8a04bcb67d8c324808ec)

carljm commented on Lib/test/test_iter.py:
```diff
@@ -238,6 +244,52 @@ def test_mutating_seq_class_exhausted_iter(self):
+                    list(it)
+                    return other == "iter"
+
+            del builtins["iter"]
```
It might be worth a comment here clarifying that we have to first `del` because otherwise the next line won't replace the key in the dict; otherwise it would be easy for someone to unthinkingly remove this `del` as a no-op

ionite34 replied:
Added some additional comments for this and other test elements in [178b8ea](https://github.com/python/cpython/pull/101769/commits/178b8ea41bb045eed63d06674e2f337d4de97bef)

carljm commented on Objects/listobject.c:
```diff
@@ -3444,18 +3444,23 @@ listiter_reduce_general(void *_it, int forward)
+            return Py_BuildValue("N(O)n", iter, it->it_seq, it->it_index);
         }
     } else {
+        PyObject *reversed = _PyEval_GetBuiltin(&_Py_ID(reversed));
```
is it worth also testing the `reversed` case for lists?

ionite34 replied:
Added case for this in [49ba8c3](https://github.com/python/cpython/pull/101769/commits/49ba8c3bfc09e026970bc4fa842c466dcee7d0f5)

carljm commented on Objects/unicodeobject.c:
```diff
@@ -14784,14 +14784,15 @@ PyDoc_STRVAR(length_hint_doc, "Private method returning an estimate of len(list(
 unicodeiter_reduce(unicodeiterobject *it, PyObject *Py_UNUSED(ignored))
 {
+    PyObject *iter = _PyEval_GetBuiltin(&_Py_ID(iter));
+
```
Any reason not to include the same comment here as in the other cases?

ionite34 replied:
Not sure, I think I originally thought this wasn't reproducible in standard conditions since strings are immutable, but it very much is since it's just a string iterator being consumed. 

Added similar comment in [93854e1](https://github.com/python/cpython/pull/101769/commits/93854e172e0b4c7cb3b999bdec6c97da74853a90)

ionite34 committed:
Refactor iter reduce builtins mutation tests

- Added some comments for dict del usages
- Switched to `__builtin__` instead of conditional `__dict__` access
- Use kwargs for improved readability

ionite34 committed:
Update iter mutating builtins test to include reversed iterator for list conditional branch

ionite34 committed:
Add comment in unicodeiter_reduce for moving iter call before it pointers

ionite34 committed:
Change test `__builtins__` to builtins import

carljm commented on Lib/test/test_iter.py:
```diff
@@ -238,6 +247,81 @@ def test_mutating_seq_class_exhausted_iter(self):
+        finally:
+            # del is required here
+            # to avoid calling our custom __eq__
+            # we also need to supress KeyErrors in case
```
```suggestion
            # we also need to suppress KeyErrors in case
```

ionite34 replied:
Changed alongside comment phrasing this was in [19ab9c6](https://github.com/python/cpython/pull/101769/commits/19ab9c63b275eaa3b6e10e04d5eec642083f233f)

carljm commented on Lib/test/test_iter.py:
```diff
@@ -238,6 +247,81 @@ def test_mutating_seq_class_exhausted_iter(self):
+            with contextlib.suppress(KeyError):
+                del builtins_dict["iter"]
+            with contextlib.suppress(KeyError):
+                del builtins_dict["reversed"]
```
I'm assuming here that it's important we do both deletions before either set? if not, we could collapse this into the existing for loop

```suggestion
            for key in orig.keys():
                with contextlib.suppress(KeyError):
                    del builtins_dict[key]
```

ionite34 replied:
Yes I think this is better behavior wise actually, they can be in the same loop, changed to
```python
for key, func in orig.items():
    with contextlib.suppress(KeyError):
        del builtins_dict[key]
    builtins_dict[key] = func
```
in [19ab9c6](https://github.com/python/cpython/pull/101769/commits/19ab9c63b275eaa3b6e10e04d5eec642083f233f)

carljm commented on Lib/test/test_iter.py:
```diff
@@ -238,6 +247,81 @@ def test_mutating_seq_class_exhausted_iter(self):
+
+        def run(builtin_name, item, init=None, sentinel=None):
+            if init is not None:
+                args = init if isinstance(init, tuple) else (init,)
```
nit, but this seems like unnecessary magic for a test case; i'm not sure it really increases readability/clarity of the cases vs just letting them all be literal tuples, mostly of length 1, given that it means you have to understand that `()` means something very different from any other single value

ionite34 replied:
Yeah, in retrospect I don't think we need to construct the actual object within the function anyways, I've simplified it to just
```python
def run(builtin_name, item, sentinel=None):
    it = iter(item) if sentinel is None else iter(item, sentinel)
    ...
```
in [19ab9c6](https://github.com/python/cpython/pull/101769/commits/19ab9c63b275eaa3b6e10e04d5eec642083f233f)

carljm commented on Misc/NEWS.d/next/Core and Builtins/2023-02-10-07-21-47.gh-issue-101765.MO5LlC.rst:
```diff
@@ -0,0 +1 @@
+Fix SystemError / segmentation fault in iter ``__reduce__`` when calling ``__builtins__.__dict__["iter"]`` mutates the iter object.
```
nit: "calling" is an unusual term for subscripting a dictionary; I think it could just be eliminated here.
```suggestion
Fix SystemError / segmentation fault in iter ``__reduce__`` when ``__builtins__.__dict__["iter"]`` mutates the iter object.
```

ionite34 replied:
Yeah I agree, also I think I'll remove the `iter` key specifically since this now affects `reversed` as well. How about

```
Fix SystemError / segmentation fault in iter ``__reduce__`` when internal access of ``builtins.__dict__`` keys mutates the iter object.
```

carljm reviewed (approved):
Just a few minor nits, but otherwise this LGTM.

I assume you did verify that the added test fails/crashes if any of the code changes are reverted? (I guess assuming you are on a system where arg eval order is such that this crash does repro.)

ionite34 committed:
Change NEWS blurb phrasing

ionite34 committed:
Update iter reduce mutating builtins test comments and simplify logic

JelleZijlstra commented on Lib/test/test_iter.py:
```diff
@@ -238,6 +247,75 @@ def test_mutating_seq_class_exhausted_iter(self):
+            # a failed test deletes the key without setting anything
+            for key, func in orig.items():
+                with contextlib.suppress(KeyError):
+                    del builtins_dict[key]
```
Why do we need to del first? We set the correct key on the next line anyway.

ionite34 replied:
It's to not invoke the `__eq__` from our test `CustomStr`, since the string key will collide with the same hash. I added some more comments regarding this in [c67b11a](https://github.com/python/cpython/pull/101769/commits/c67b11ac520047e63e0ca9fcd5d13cba96491f3f)

JelleZijlstra replied:
Ah, right! This thing is breaking my assumptions about how dicts should behave :)

JelleZijlstra commented on Objects/bytearrayobject.c:
```diff
@@ -2391,11 +2391,17 @@ PyDoc_STRVAR(length_hint_doc,
+
+    /* _PyEval_GetBuiltin can invoke arbitrary code.
+     * calls must be *before* access of `it` pointers,
+     * since C parameter eval order is undefined.
```
I think parameter eval order isn't the issue and shouldn't be mentioned. The problem is that the `_PyEval_GetBuiltin` call invalidates the previous `if` check.

For example, this code would also be buggy:
```
    if (it->it_seq != NULL) {
        PyObject *iter = _PyEval_GetBuiltin(&_Py_ID(iter));
        return Py_BuildValue("N(O)n", iter, it->it_seq, it->it_index);
    }
```

The parameter evaluation order makes things a bit more unpredictable, but the bug is that the `if` check is done too early, not that we're relying on evaluation order.

ionite34 replied:
Hm, yeah. I think this might be better?
```
    /* _PyEval_GetBuiltin can invoke arbitrary code,
     * call must be before access of iterator pointers.
     * see issue #101765 */
```

JelleZijlstra replied:
sounds good!

ionite34 replied:
Updated all comments to the above in [9b664c2](https://github.com/python/cpython/pull/101769/commits/9b664c225c44e14870b83c59a003d7a879a99831)

JelleZijlstra reviewed (commented):
Sorry for dropping this for a while! I have a few minor pieces of feedback.

ionite34 committed:
Update comments to better reflect issue

ionite34 committed:
Clarify test comments

kumaraditya303 commented on Objects/listobject.c:
```diff
@@ -3444,18 +3444,22 @@ listiter_reduce_general(void *_it, int forward)
     } else {
+        PyObject *reversed = _PyEval_GetBuiltin(&_Py_ID(reversed));
         listreviterobject *it = (listreviterobject *)_it;
         if (it->it_seq) {
```
If `it_seq` is NULL then reversed is leaked here, `_PyEval_GetBuiltin` returns a strong reference.

JelleZijlstra replied:
Good point, but doesn't that mean the old code also leaked references? I'll prepare a PR to fix this now.

JelleZijlstra replied:
Oh wait, the N code to Py_BuildValue steals a reference, so the previous code was right in terms of refcounting.
